### PR TITLE
add line-range

### PR DESCRIPTION
**Pull request by tskinn**
This is my first attempt at actually writing some rust code so bear with me.

I wasn't sure of the best place to put the logic for checking line range. I started with putting it in the print_line function but moved it out to be able to short circuit and stop processing once out of the range.

I didn't add the ability to do multiple ranges yet but it should be fairly simple.

How robust of error checking do we need on validating the line ranges? Like do we want to sort or make sure they are not conflicting or that they are in least to greatest order or anything like that?

### CODE REVIEW

**Comment by tskinn** [comment]
I appreciate the feedback. I will start working on this. Thanks for you patience. Feel free to let me know if you think I'm going in the completely wrong direction with this.

File: src/app.rs [hunk 1]
```diff
@@ -322,10 +336,69 @@ pub struct Config<'a> {
...
         .unwrap_or(false)
 }
+
+pub struct LineRange {
+    pub lower: usize,
+    pub upper: usize,
+}
+
+impl LineRange {
+    pub fn from(range_raw: &str) -> LineRange {
```
Comment (sharkdp) [inline src/app.rs]: I think this should also return a `Result<LineRange>` (by removing the `unwrap_or` call below). This way, we can pass the error messages through to the call-site.

File: src/app.rs [hunk 2]
```diff
@@ -322,10 +336,69 @@ pub struct Config<'a> {
...
+        LineRange{
+            lower: usize::min_value(),
+            upper: usize::max_value(),
+        }
+    }
+
+    pub fn parse_range(range_raw: &str) -> Result<LineRange> {
+        let mut new_range = LineRange::new();
+
+        if range_raw.bytes().nth(0).ok_or("No first byte")? == b':' {
```
Comment (sharkdp) [inline src/app.rs]: Maybe the error message here (and below) should be "Empty line range"?

File: src/app.rs [hunk 3]
```diff
@@ -322,10 +336,69 @@ pub struct Config<'a> {
...
+            new_range.upper = range_raw[1..].parse()?;
+            return Ok(new_range);
+        } else if range_raw.bytes().last().ok_or("No last byte")? == b':' {
+            new_range.lower = range_raw[..range_raw.len() - 1].parse()?;
+            return Ok(new_range);
+        }
+
+        let line_numbers: Vec<&str> = range_raw.split(':').collect();
+        if line_numbers.len() == 2 {
```
Comment (sharkdp) [inline src/app.rs]: I think we should have an `else` branch here which returns a error ("expected single ':' character").

File: src/app.rs [hunk 4]
```diff
@@ -322,10 +336,69 @@ pub struct Config<'a> {
...
+            new_range.lower = line_numbers[0].parse()?;
+            new_range.upper = line_numbers[1].parse()?;
+        }
+        Ok(new_range)                
+    }
+}
+
+#[test]
+fn test_parse_line_range_full() {
+    let range = LineRange::from(Some("40:50")).expect("Shouldn't fail on test!");
```
Comment (sharkdp) [inline src/app.rs]: This would turn into `::from("40:50")`.

File: src/app.rs [hunk 5]
```diff
@@ -322,10 +336,69 @@ pub struct Config<'a> {
...
+    assert_eq!(usize::min_value(), range.lower);
+    assert_eq!(50, range.upper);
+}
+
+#[test]
+fn test_parse_line_range_partial_max() {
+    let range = LineRange::from(Some("40:")).expect("Shouldn't fail on test!");
+    assert_eq!(40, range.lower);
+    assert_eq!(usize::max_value(), range.upper);
+}
```
Comment (sharkdp) [inline src/app.rs]: If we change the signature of `LineRange::from` to return a `Result`, we can also add some tests for negative cases.

File: src/features.rs [hunk 6]
```diff
@@ -85,28 +85,72 @@ fn print_file(
...
+        printer.print_footer()?;
+    }
+    Ok(())
+}
 
-    let mut reader: Box<BufRead> = match filename {
-        None => Box::new(stdin.lock()),
-        Some(filename) => Box::new(BufReader::new(File::open(filename)?)),
-    };
+fn print_file_ranges<'a>(
```
Comment (sharkdp) [inline src/features.rs]: It would be really great if we had just a single function and would avoid this code duplication. I would try to change the `ranges` parameter to accept an `Option<Range>`.

File: src/main.rs [hunk 7]
```diff
@@ -41,6 +41,7 @@ mod errors {
             Clap(::clap::Error);
             Io(::std::io::Error);
             SyntectError(::syntect::LoadingError);
+            ParseIntError(::std::num::ParseIntError);
```
Comment (sharkdp) [inline src/main.rs]: Nice!

File: src/app.rs [hunk 8]
```diff
@@ -269,6 +282,7 @@ impl App {
             term_width: Term::stdout().size().1 as usize,
             files,
             theme: self.matches.value_of("theme"),
+            line_range: self.matches.value_of("line-range").map(LineRange::from),
```
Comment (sharkdp) [inline src/app.rs]: If we change the signature of `LineRange::from`, as suggested below, we could use
``` rust
line_range: tranpose(self.matches.value_of("line-range").map(LineRange::from))
```
such that the parsing-error-messages would bubble up to the main function (and reported to the user).

Here, `transpose` is a little helper function that turns a "Option of Result" into a "Result of Option". It could be implemented like this:
```
fn transpose<T, E>(opt: Option<Result<T, E>>) -> Result<Option<T>, E> {
    opt.map_or(Ok(None), |res| res.map(Some))
}
```

This function is actually an experimental feature which means that we could soon replace this by a call to the standard-library `.transpose()`: https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.transpose

File: src/app.rs (hunk 3 above)
Comment (tskinn) [reply src/app.rs]: This might be a silly question but how do I go about creating an error? When I try to return
```Err("expected single....")``` I get a mismatched types error: ```expected struct `errors::Error`, found reference```

Do I need to implement a new error type or something like that?

File: src/app.rs (hunk 4 above)
Comment (tskinn) [reply src/app.rs]: Does this only work if the tests are in a new module called LineRange?

File: src/app.rs (hunk 8 above)
Comment (tskinn) [reply src/app.rs]: OK I think I see what you are saying now. We want to bubble up any errors that happen when parsing the linerange to bubble up from the config() function.

File: src/app.rs (hunk 3 above)
Comment (sharkdp) [reply src/app.rs]: You should be able to use `Err("expected single....".into())`.

File: src/app.rs (hunk 4 above)
Comment (sharkdp) [reply src/app.rs]: Ah, sorry. I just meant to say that the `Some` part will not be needed anymore. I would still use `LineRange::from`.

Comment (tskinn) [reply src/app.rs]: Oh duh. Gotcha
